include/acl_thread.h

```diff
@@ -105,4 +116,54 @@ void acl_yield_lock_and_thread();
 } /* extern "C" */
 #endif
 
+// -- RAII wrapper classes --
+// In order to enable RAII practices, we need to construct some mutex classes
+// and RAII wrapper. The mutex class can interact with std::lock_guard. Note
+// that std::scoped_lock could not be adapted since acl_threadsupport does not
+// provide the try_lock() functionality.
+
+// Mutex class wrapper around acl_lock() and acl_unlock().
+// This meets the BasicLockable requirement for std::lock_guard.
+class acl_mutex_wrapper_t {
+public:
+  void lock() { acl_lock(); }
+  void unlock() { acl_unlock(); }
+};
+
+// Custom implementation of std::lock_guard that takes in a boolean argument to
+// determine if to lock or not. Future improvement: this should be templated to
+// take in any mutex class that adheres to the BasicLockable requirement.
+class acl_condition_lock_guard {
+public:
+  explicit acl_condition_lock_guard(acl_mutex_wrapper_t &mtex, bool cond)
+      : mtex(mtex), cond(cond) {
+    if (cond) {
+      mtex.lock();
+    }
+  }
+  ~acl_condition_lock_guard() {
+    if (cond) {
+      mtex.unlock();
+    }
+  }
+
+private:
+  acl_mutex_wrapper_t &mtex;
+  bool cond;
+};
```

Comment:
Does the deferrable lock in unique_lock always unlocks when it destructs? If so, then it will mess with the underlying mechanism with acl_lock() and acl_unlock(). This introduces more risks. We could try this but it will require more testing...I don't have bandwidth to do anymore testing. 

I would vote for just using my acl_condition_lock_gaurd class. It also mimics